src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -0,0 +1,440 @@
+      stats.addIntValue(bigIntegerVal);
+      bytes = bigIntegerVal.toByteArray();
+      if (bytes.length < length) {
+        byte[] newBytes = new byte[length];
+        Arrays.fill(newBytes, (byte) (bytes[0] < 0 ? -1 : 0));
+        System.arraycopy(bytes, 0, newBytes, length - bytes.length, bytes.length);
+        bytes = newBytes;
+      }
+    }
+    if (bytes != null) {
+      checkFixedLengthByteArray(bytes, length, insertRowsCurrIndex);
+      return bytes;
+    }
+    throw new SFException(
+        ErrorCode.UNKNOWN_DATA_TYPE, logicalTypeAnnotation, type.getPrimitiveTypeName());
+  }
+
+  /**
+   * Parses a group value based on Parquet group logical type.
+   *
+   * @param value value to parse
+   * @param type Parquet column type
+   * @param stats column stats to update
+   * @param defaultTimezone default timezone to use for timestamp parsing
+   * @param insertRowsCurrIndex Used for logging the row of index given in insertRows API
+   * @return list of parsed values
+   */
+  private static ParquetBufferValue getGroupValue(
+      Object value,
+      GroupType type,
+      RowBufferStats stats,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex) {
+    LogicalTypeAnnotation logicalTypeAnnotation = type.getLogicalTypeAnnotation();
+    if (logicalTypeAnnotation == null) {
+      return getStructValue(value, type, stats, defaultTimezone, insertRowsCurrIndex);
+    }
+    if (logicalTypeAnnotation instanceof ListLogicalTypeAnnotation) {
+      return get3LevelListValue(value, type, stats, defaultTimezone, insertRowsCurrIndex);
+    }
+    if (logicalTypeAnnotation instanceof MapLogicalTypeAnnotation) {
+      return get3LevelMapValue(value, type, stats, defaultTimezone, insertRowsCurrIndex);
+    }
+    throw new SFException(
+        ErrorCode.UNKNOWN_DATA_TYPE, logicalTypeAnnotation, type.getClass().getSimpleName());
+  }
+
+  /**
+   * Parses a struct value based on Parquet group logical type. The parsed value is a list of
+   * values, where each element represents a field in the group. For example, an input {@code
+   * {"field1": 1, "field2": 2}} will be parsed as {@code [1, 2]}.
+   */
+  private static ParquetBufferValue getStructValue(
+      Object value,
+      GroupType type,
+      RowBufferStats stats,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex) {
+    Map<String, Object> structVal =
+        DataValidationUtil.validateAndParseIcebergStruct(
+            type.getName(), value, insertRowsCurrIndex);
+    List<Object> listVal = new ArrayList<>(type.getFieldCount());
+    float estimatedParquetSize = ParquetBufferValue.DEFINITION_LEVEL_ENCODING_BYTE_LEN;
+    for (int i = 0; i < type.getFieldCount(); i++) {
+      ParquetBufferValue parsedValue =
+          parseColumnValueToParquet(
+              structVal.getOrDefault(type.getFieldName(i), null),
+              type.getType(i),
+              stats,
```

Comment:
we need the nested object's "metadata" to also be passed in for stuff like precision/scale/byte[] maxlength/etc.
However we won't have a ColumnMetadata object in that situation, which means we now need an abstraction over ColumnMetadata for use in the value parsers (I'm leaving another comment soon that we should look to have one ParquetValueParser class as there's too many similarities for them to be completely independent classes. Especially since we need to support structured data types for FDN types too some day)